### ISSUE TITLE
How to access list of acceptable CAs in client authentication?

### ISSUE THREAD

**Issue by fschoenm** [randombit/botan#1261]
We implement a TLS application that requires support for client authentication. When the server sends its *CertificateRequest* to the client, it includes a list of trusted CAs for client certificates. I noticed that this list of CAs is stored in the client's handshake state (``state.cert_req().acceptable_CAs()``) but never accessed anywhere.

* The ``Credentials_Manager::cert_chain()`` function gets only told the valid key types but not the valid CAs.
* There seems to be no way to access the TLS handshake state, or the certificate request in particular, from the application.

Is that an interface oversight or was it done on purpose? Or did I miss some way to access this information from the application?

**Comment by randombit** [comment]
You are right this is an oversight in the API. I think we'll be able to support this in a future release. Thanks for raising the issue.

The information is accessible by inspecting the handshake messages directly. It's a hack but it should work. To do this, override `TLS::Callbacks::tls_inspect_handshake_msg`, then watch for the certificate request, roughly 

```
if(msg.type() == TLS::CERTIFICATE_REQUEST)
  {
  TLS::Certificate_Req& req = dynamic_cast<TLS::Certificate_Req&>(msg);
  std::vector<X509_DN> names = req.acceptable_CAs();
  // provide names to your credentials manager somehow
  }
```

This introduces an unfortunate coupling between the code that handles credentials vs the session-specific logic of the other callbacks, so not claiming this is optimal but it should at least work.

**Comment by randombit** [comment]
@fschoenm PR #1395 added support for a new callback in `Credentials_Manager` which accepts the `X509_DN` list provided by the server. This will be available starting in Botan 2.5
